core/src/main/java/com/digitalpebble/stormcrawler/protocol/file/FileResponse.java

```diff
@@ -0,0 +1,182 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.digitalpebble.stormcrawler.protocol.file;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.invoke.MethodHandles;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLDecoder;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Locale;
+
+import org.apache.http.HttpStatus;
+import org.slf4j.LoggerFactory;
+
+import com.digitalpebble.stormcrawler.Metadata;
+import com.digitalpebble.stormcrawler.protocol.HttpHeaders;
+import com.digitalpebble.stormcrawler.protocol.ProtocolResponse;
+
+public class FileResponse {
+
+    /**
+     * Adopted from Apache Nutch File Protocol implementation
+     * 
+     * @see https://github.com/apache/nutch/blob/master/src/plugin/protocol-file
+     *      /src/java/org/apache/nutch/protocol/file/FileResponse.java
+     * @see https://github.com/apache/nutch/blob/master/src/plugin/protocol-file
+     *      /src/java/org/apache/nutch/protocol/file/File.java
+     * 
+     *      give the url as "file://C:/folder/" to the seed.
+     **/
+
+    private byte[] content;
+    private int statusCode;
+    private Metadata metadata;
+    private final FileProtocol fileProtocol;
+    static final SimpleDateFormat dateFormat = new SimpleDateFormat(
+            "EEE, dd MMM yyyy HH:mm:ss zzz", Locale.US);
+    static final org.slf4j.Logger LOG = LoggerFactory.getLogger(MethodHandles
+            .lookup().lookupClass());
+
+    public FileResponse(String u, Metadata md, FileProtocol fp)
+            throws MalformedURLException, IOException {
+
+        fileProtocol = fp;
+        metadata = md;
+
+        URL url = new URL(u);
+
+        if (!url.getPath().equals(url.getFile())) {
+            LOG.warn("url.getPath() != url.getFile(): {}.", url);
+        }
+
+        String path = "".equals(url.getPath()) ? "/" : url.getPath();
+
+        File file = new File(
+                URLDecoder.decode(path, fileProtocol.getEncoding()));
+
+        if (!file.exists()) {
+            statusCode = HttpStatus.SC_NOT_FOUND;
+        }
+
+        if (!file.canRead()) {
+            statusCode = HttpStatus.SC_UNAUTHORIZED;
+        }
+
+        if (!file.equals(file.getCanonicalFile())) {
+            metadata.setValue(HttpHeaders.LOCATION, file.getCanonicalFile()
+                    .toURI().toURL().toString());
+            statusCode = HttpStatus.SC_MULTIPLE_CHOICES;
+        }
+
+        if (file.isDirectory()) {
+            getDirAsHttpResponse(file);
+        } else if (file.isFile()) {
+            getFileAsHttpResponse(file);
+        } else {
+            statusCode = HttpStatus.SC_INTERNAL_SERVER_ERROR;
+        }
+    }
+
+    public ProtocolResponse toProtocolResponse() throws IOException {
+        return new ProtocolResponse(content, statusCode, metadata);

```

Comment:
check that content is not null - which would be the case when a file does not exist and set content to an empty array instead. Otherwise we get a NPE exception in the fetcher which is unnecessary